### ISSUE TITLE
[Question] Issue with seeding.

### ISSUE THREAD

**Issue by kad99kev** [ugr-sail/sinergym#368]
Hello!

Thank you for the constant updates on the library!

I recently updated the Sinergym version (from 2.5.2 to 3.0.6) and I was just wondering if you have noticed any issue with reproducibility. Earlier, if I performed training or evaluation with seeds, I would get same results when I would run the same model or hyperparameters. However, now I get a difference in performance and output whenever I run the same hyperparameter or model. I believe this is due to the normalisation wrapper.

Do you have any suggestions on how I can ensure reproducibility?

Thank you for your time!

**Comment by AlejandroCN7** [comment]
Hi @kad99kev!

Reproducibility is something that should be maintained. Unfortunately, between the versions you mention, it is impossible because the entire backend of the tool has been updated. That is, the same experiment with the same seed will not give the same result between those two versions (but it should in the same version).

However, you comment that you get a difference in performance and output with the same model and seed (in version 3.0.6), right? I hold this issue open to check later when I have time. Pointing out the normalization wrapper is a great idea, thank you very much. This wrapper was updated based on [this Gymnasium wrapper](https://gymnasium.farama.org/api/wrappers/observation_wrappers/#gymnasium.wrappers.NormalizeObservation) recently in order to get normalizations while the environment write observations (and not have to calculate ranges that are not provided by EnergyPlus). Perhaps this process affects the randomness of the experiment and causes the seeds to malfunction. As I said, I will check it to find the problem.

Thank you very much for writing this issue,
Regards!

**Comment by kad99kev** [comment]
Hi @AlejandroCN7,

Yes, to confirm, when I run the same model and the same seed multiple times in 2.5.2, I observe reproducibility. For the same model and the same seed, when run multiple times in 3.0.6, each time would give a near (sometimes different) result.

Thank you for checking it out. Do let me know if you are able to find the root cause. As I mentioned, I feel the normalisation wrapper may introduce floating point errors, which pass through multiple iterations, causing major divergence. But I still believe if it was seeded correctly, this floating issue should not come up. Do let me know what you think and if I can help in any way.

Thank you once again!

**Comment by AlejandroCN7** [comment]
Hi @kad99kev!

Can you try doing a run without the normalization wrapper (as simple as you can)? It is to check if definitely the error is caused by this wrapper. If not, don't worry, it's just that this week has been a bit dense for me and I haven't been able to check it yet.

Thanks to you!

**Comment by luckypipitao** [comment]
Hello!

Thank you for the constant updates on the library!

I recently encountered a problem when using sinergym's Eplus-datacenter-mixed-continuous-v1 environment. The first point is that the environment only has 2 action spaces while the web page displays 4 action spaces. Second point, I found that these two action spaces only control west_zone_temperature but not east_zone_temperature.
There is another environment where a similar problem occurs in the Eplus-5zone-mixed-discrete-v1 environment. The dates I used are (1,8,1991,20,9,1991). Even in the summer, indoors When the temperature reaches 30 degrees, the power of the VAV system is also zero, even if I set the clg_setting point very low.

Is there a way to convert the action space of Eplus-datacenter-mixed-continuous-v1 into a 4-dimensional one, or solve the problem that the VAV exceeds the clg setting value and does not work?

Finally, thank you for creating this simulation platform.

Thank you very much for writing this issue,
Regards!

**Comment by AlejandroCN7** [comment]
Hi @luckypipitao!

Thank you for your work. As this is a different topic from the issue, I would be grateful if you could ask your questions in a new issue for the next one if it is not a problem :)

Going back to your questions, it doesn't work exactly like that in the case of datacenter. Although the temperature setpoints are observable in both zones separately as you indicate, the schedule that manages it is shared by both zones. That is, by setting the heating and cooling values to a specific value, the setpoints of both zones are set to exactly the same value.

This can be observed in monitor.csv if you have the logger active, as when you set a value these are duplicated for observation in the next timestep. If you want to control each zone independently, it requires modifying the building model manually (creating a new schedule, modifying thermostats, etc).

Moving on to the 5zone building, we have not had that problem, I will look into it in any case.

Regards!

**Comment by kad99kev** [comment]
Hello, sorry for the late update.

I ran the code below:
```python
import gymnasium as gym
import numpy as np
from gymnasium.wrappers.normalize import NormalizeReward
import random
import numpy as np
import torch

import sinergym
from sinergym.utils.wrappers import (LoggerWrapper,
                                     NormalizeObservation)

# TRY NOT TO MODIFY: seeding
random.seed(0)
np.random.seed(0)
torch.manual_seed(0)

env = gym.make('Eplus-5zone-hot-continuous-stochastic-v1')
# env = NormalizeAction(env)
# env = NormalizeObservation(env)
env = NormalizeReward(env)
env = LoggerWrapper(env)
env.action_space.seed(0)

for i in range(1):
    obs, info = env.reset(seed=0)
    rewards = []
    terminated = False
    current_month = 0
    j = 0
    while not terminated:
        a = env.action_space.sample()
        obs, reward, terminated, truncated, info = env.step(a)
        print(a, obs[:5], reward)
        if j > 2:
            raise
        j += 1
        # rewards.append(reward)
        # if info['month'] != current_month:  # display results every month
        #     current_month = info['month']
        #     print('Reward: ', sum(rewards), info)
    # print(
    #     'Episode ',
    #     i,
    #     'Mean reward: ',
    #     np.mean(rewards),
    #     'Cumulative reward: ',
    #     sum(rewards))
env.close()
```

And I get the following output:

```bash
# With Normalisation
[ 0.27392337 -0.46042657] [ 0.00499969  0.00499969  0.          0.95822201 -0.97456266] -81.38322082320886
[-0.918053   -0.96694475] [ 0.00408214  0.00408214  0.          1.21089681 -1.2170853 ] -2.0189386477117317
[0.6265405  0.82551116] [ 0.00353516  0.00353516  0.          1.33532966 -1.33841824] -1.2630430770563694
[0.21327156 0.4589931 ] [ 0.00316187  0.00316187  1.99989995  0.93786803 -0.85638926] -0.578626278160322

# For No Normalisation
[ 0.27392337 -0.46042657] [ 1.        1.        0.        5.128147 61.      ] -81.38322082320886
[-0.918053   -0.96694475] [ 1.         1.         0.         5.3640738 57.       ] -2.0189386477117317
[0.6265405  0.82551116] [ 1.   1.   0.   5.6 53. ] -0.845193831797655
[0.21327156 0.4589931 ] [ 1.         1.         1.         5.5607243 54.25     ] -0.5861029288783863
```

Would you know why there is a difference when the same actions are performed? Is there something I am missing while seeding?

**Comment by AlejandroCN7** [comment]
Hi @kad99kev! 

I think I have found the solution to the problem as shown in the PR I have created. I copy here the solution to make it clear :)

## Solution

I have updated the action process callback in backend simulator. Specifically, I have replaced the `callback_after_predictor_after_hvac_managers` callback by `callback_begin_system_timestep_before_predictor`.

This solves the problem, but there is another problem. Without knowing the origin, sometimes the simulator starts to apply real-time control on the second timestep, and sometimes on the third timestep. This causes the chain of events to differ even though the seeding is the same. In other words, it is as if "two scenarios" could occur for the same seed, as shown in the following screenshots.

![seeding](https://github.com/ugr-sail/sinergym/assets/23475034/867e1c91-aae9-490f-8736-63f39daffd93)

For the case of 5zones, the values 12.8 and 40.0 in the setpoints are the default values until the EnergyPlus API can start controlling. This image shows the two possible versions: one where it starts controlling at the third timestep and one where it starts controlling at the fourth timestep with a timestep delay. I don't understand what is the origin of this problem, but it seems to be from EnergyPlus itself.

In summary, this change makes reproducibility of results possible, but it is necessary to check that it starts to control in the same timestep to ensure reproducibility.
